File: python/Q121_best_time_to_buy_and_sell_stock.py
#
# @lc app=leetcode id=121 lang=python3
#
# [121] Best Time to Buy and Sell Stock
#


# @lc code=start
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        left: int = 0
        profit: int = 0

        for right in range(len(prices)):
            if prices[left] > prices[right]:
                left = right
            elif prices[right] - prices[left] > profit:
                profit = prices[right] - prices[left]

        return profit


# @lc code=end

File: rust/src/leetcode/Q121_best_time_to_buy_and_sell_stock.rs
/*
 * @lc app=leetcode id=121 lang=rust
 *
 * [121] Best Time to Buy and Sell Stock
 */

// @lc code=start
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut min_buy: i32 = prices[0];
        let mut profit: i32 = 0;

        for price in prices {
            if price - min_buy > profit {
                profit = price - min_buy;
            }

            if price < min_buy {
                min_buy = price;
            }
        }

        profit
    }
}
// @lc code=end
